network/p2p/throttled_handler.go

```diff
@@ -0,0 +1,39 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package p2p
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/ava-labs/avalanchego/ids"
+)
+
+var (
+	ErrThrottled         = errors.New("throttled")
+	_            Handler = (*ThrottledHandler)(nil)
+)
+
+type ThrottledHandler struct {
+	Handler
+	Throttler Throttler
+}
+
+func (t ThrottledHandler) AppGossip(ctx context.Context, nodeID ids.NodeID, gossipBytes []byte) error {
+	if !t.Throttler.Handle(nodeID) {
+		return fmt.Errorf("dropping message from %s: %w", nodeID, ErrThrottled)
```

Comment:
q: should we really return an error when throttling? How about logging a warn and return nil. Not sure how AppGossip errors are handled by consensus engine

Reply:
This is dropped and logged by the `p2p.responder` that wraps all `p2p.Handler` instances. Currently we log, but in the future we'll implement application-level errors in https://github.com/ava-labs/avalanchego/issues/1889 where we'll make a custom `p2p.AppError` type that the sdk will interpret as an `error`.